include/audio_backend/__asio_audio_device.h

```diff
@@ -0,0 +1,527 @@
+
+    if (ERROR_SUCCESS != result) {
+      throw runtime_error("Failed to read info from registry: 0x" + result);
+    }
+  }
+
+  ~__reg_key_reader() {
+    if (_key) {
+      RegCloseKey(_key);
+    }
+  }
+
+  vector<string> subkeys() const {
+    constexpr int max_asio_name_length{32};
+    char name[max_asio_name_length];
+    DWORD size{max_asio_name_length};
+    vector<string> keys;
+
+    for (DWORD index = 0; ; ++index) {
+      const auto result = RegEnumKeyEx(_key, index, name, &size, nullptr, nullptr, nullptr, nullptr);
+
+      if (result != ERROR_SUCCESS) {
+        break;
+      }
+
+      keys.emplace_back(name);
+    }
+
+    return keys;
+  }
+
+  __reg_key_reader subkey(const string& name)
+  {
+    return {_key, name};
+  }
+
+  string value(const char* name) const {
+    DWORD size{0};
+    RegQueryValueExA(_key, name, nullptr, nullptr, nullptr, &size);
+
+    string value(size + 1, 0);
+    RegQueryValueExA(_key, name, nullptr, nullptr, LPBYTE(value.data()), &size);
+
+    return value;
+  }
+
+private:
+  HKEY _key{0};
+};
+
+class __asio_devices {
+public:
+  static __asio_devices& get() {
+    static __asio_devices instance;
+    return instance;
+  }
+
+  ~__asio_devices() {
+    CoUninitialize();
+  }
+
+  optional<audio_device> default_input() {
+    auto devices = inputs();
+    if (devices.empty()) {
+      return {};
+    }
+    return *devices.begin();
+  }
+
+  optional<audio_device> default_output() {
+    auto devices = outputs();
+    if (devices.empty()) {
+      return {};
+    }
+    return *devices.begin();
+  }
+
+  audio_device_list inputs() {
+    return enumerate(audio_direction::in);
+  }
+
+  audio_device_list outputs() {
+    return enumerate(audio_direction::out);
+  }
+
+private:
+  __asio_devices() {
+    CoInitialize(nullptr);
+  }
+
+  audio_device_list enumerate(audio_direction direction) {
+    audio_device_list devices;
+
+    __reg_key_reader asio_reg(HKEY_LOCAL_MACHINE, "software\\asio");
+
+    int index = 0;
+    for (const auto& name : asio_reg.subkeys()) {
+
+      if (is_excluded(name)) {
+        continue;
+      }
+      const auto key = asio_reg.subkey(name);
+      const auto clsid = key.value("CLSID");
+
+      CLSID class_id;
+      CLSIDFromString(CComBSTR(clsid.c_str()), &class_id);
+
+      auto device = audio_device{index++, name, class_id, direction};
+      if (is_required(device, direction)) {
+        devices.push_front(move(device));
+      }
+    }
+    return devices;
+  }
+
+  static bool is_required(const audio_device& device, const audio_direction direction) {
+    switch (direction) {
+    case audio_direction::in: return device.is_input();
+    case audio_direction::out: return device.is_output();
+    case audio_direction::full_duplex: return device.is_input() && device.is_output();
+    default: return false;
+    };
+  }
+
+  bool is_excluded(string_view name)
+  {
+    if (name == "Realtek ASIO") {
```

Comment:
Please add a comment indicating why Realtek ASIO is excluded.

Reply:
Added comment in 4d7ae72

My first attempt to support Realtek failed. I'm assuming that I mishandled something, but it would oscillate between successfully initialising on one pass, failing to initialise on the next. There was also some interaction with the OS, I got pop-ups from my security tool asking for permission to stream audio. I will come back to this in a future commit.

Reply:
Realtek seems unreliable. Works perfectly on one run, fails with ASE_InvalidMode & no error string on the next run. Flips between these two states forever